runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,484 @@
+  }
+
+  // If prev is NULL then we have the actual head of the list
+  if (prev == NULL) {
+    node->deferList = NULL;
+    return dnode;
+  }
+
+  // If neither are NULL, we need to correct list so previous points to NULL
+  prev->next = NULL;
+  return dnode;
+}
+
+// Requires spinlock.
+static inline struct defer_node *pop_all_defer_list(struct tls_node *node);
+static inline struct defer_node *pop_all_defer_list(struct tls_node *node) {
+  struct defer_node *head = node->deferList;
+  node->deferList = NULL;
+  return head;
+}
+
+// Requires spinlock.
+static inline void push_defer_list(struct tls_node *node, struct defer_node *dnode);
+static inline void push_defer_list(struct tls_node *node, struct defer_node *dnode) {
+  dnode->next = node->deferList;
+  node->deferList = dnode;
+}
+
+// Must be called from the thread owning the node.
+static inline struct defer_node *pop_recycle_list(struct tls_node *node);
+static inline struct defer_node *pop_recycle_list(struct tls_node *node) {
+  struct defer_node *dnode = node->recycleList;
+  
+  // If empty, return fresh allocation
+  if (dnode == NULL) {
+    return chpl_mem_calloc(1, sizeof(struct defer_node), 0, 0, 0);
+  }
+
+  node->recycleList = dnode->next;
+
+  // Sanitize before return...
+  memset(dnode, 0, sizeof(struct defer_node));
+  return dnode;
+}
+
+// Must be called from the thread owning the node.
+static inline void push_recycle_list(struct tls_node *node, struct defer_node *dnode);
+static inline void push_recycle_list(struct tls_node *node, struct defer_node *dnode) {
+  dnode->next = node->recycleList;
+  node->recycleList = dnode;
+}
+
+// Deletes data held by 'dnode', does not free 'dnode'
+static inline void delete_data(struct defer_node *dnode);
+static inline void delete_data(struct defer_node *dnode) {
+  if (dnode->numData) {
+      void **arr = dnode->data;
+      for (int i = 0; i < dnode->numData; i++) {
+          chpl_mem_free(arr[i], 0, 0);
+      }
+  }
+  chpl_mem_free(dnode->data, 0, 0);
+}
+
+// Initializes TLS; should only need to be called once.
+static void init_tls(void);
+static void init_tls(void) {
+  struct tls_node *node = chpl_calloc(1, sizeof(struct tls_node));
+  node->epoch = get_global_epoch();
+
+  // Append to head of list
+  struct tls_node *old_head;
+  do {
+    old_head = chpl_qsbr_tls_list;
+    node->next = old_head;
+  } while (!atomic_compare_exchange_weak_uintptr_t(
+      (uintptr_t *) &chpl_qsbr_tls_list, (uintptr_t) old_head, (uintptr_t) node)
+    );
+
+  CHPL_TLS_SET(chpl_qsbr_tls, node);
+}
+
+// Obtains the minimum epoch of all threads.
+static uint64_t safe_epoch(void);
+static uint64_t safe_epoch(void) {
+  uint64_t min = (uint64_t) -1;
+  // Check all remaining threads for whether they have updated to a more recent epoch.
+  for (struct tls_node *node = get_tls_list(); node != NULL; node = node->next) {
+    if (!is_parked(node)) {
+      uint64_t epoch = get_epoch(node);
+      min = min ? (min < epoch) : epoch;       
```

Comment:
Should that be ``min = (min < epoch) ? min : epoch``?

Reply:
Indeed it should, that's a critical bug I need to fix asap.